nmostesting/ControllerTest.py

```diff
@@ -0,0 +1,530 @@
+
+# asyncio queue for passing Testing Façade answer responses back to tests
+_event_loop = asyncio.new_event_loop()
+asyncio.set_event_loop(_event_loop)
+_answer_response_queue = asyncio.Queue()
+
+# use exit Event to quit tests early that involve waiting for senders/connections 
+exitTestEvent = Event()
+
+app = Flask(__name__)
+TEST_API = Blueprint('test_api', __name__)
+
+class TestingFacadeException(Exception):
+    """Exception thrown due to comms or data errors between NMOS Testing and Testing Façade"""
+    pass
+
+@TEST_API.route(CALLBACK_ENDPOINT, methods=['POST'])
+def retrieve_answer():
+
+    if request.method == 'POST':
+        if 'name' not in request.json:
+            return 'Invalid JSON received'
+
+        _event_loop.call_soon_threadsafe(_answer_response_queue.put_nowait, request.json)
+
+        # Interupt any 'sleeps' that are still active 
+        exitTestEvent.set()
+
+    return 'OK'
+
+class ControllerTest(GenericTest):
+    """
+    Testing initial set up of new test suite for controller testing
+    """
+    def __init__(self, apis, registries, node, dns_server):
+        GenericTest.__init__(self, apis)
+        self.authorization = False  # System API doesn't use auth, so don't send tokens in every request
+        self.primary_registry = registries[1]
+        self.node = node
+        self.dns_server = dns_server
+        self.mock_registry_base_url = ''
+        self.mock_node_base_url = ''
+        self.question_timeout = 600 # default timeout in seconds
+        self.test_data = self.load_resource_data()
+        self.senders = [] 
+        self.receivers = [] # receiver list containing: {'label': '', 'description': '', 'id': '', 'registered': True/False, 'connectable': True/False, 'answer_str': ''}
+        self.senders_ip_base = '239.3.14.' # Random multicast IP to assign to senders
+
+    def set_up_tests(self):
+        if self.dns_server:
+            self.dns_server.load_zone(self.apis[CONTROLLER_TEST_API_KEY]["version"], self.protocol, self.authorization,
+                                      "test_data/controller/dns_records.zone", CONFIG.PORT_BASE+100)
+            self.dns_server.set_expected_query(
+                QTYPE.PTR,
+                [
+                    "_nmos-query._tcp.{}.".format(CONFIG.DNS_DOMAIN)
+                ]
+            )
+        # Reset registry to clear previous heartbeats, etc.
+        self.primary_registry.reset()
+        self.primary_registry.enable()
+        self.mock_registry_base_url = 'http://' + get_default_ip() + ':' + str(self.primary_registry.get_data().port) + '/'
+        self.mock_node_base_url = 'http://' + get_default_ip() + ':' + str(self.node.port) + '/'
+
+        # Populate mock registry with senders and receivers and store the results
+        self._populate_registry()
+
+        # Set up mock node
+        self.node.registry_url = self.mock_registry_base_url
+
+        print('Registry should be available at ' + self.mock_registry_base_url)
```

Comment:
Just noting that `print` is used pretty sparingly in the testing tool so far. We have a written rule about what's worth logging and what format should be used, but let's add that to the list of things to review :-P

Reply:
All print statements introduced as part of this development process have been removed.